Userland/tail.cpp

```diff
@@ -0,0 +1,112 @@
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <errno.h>
+#include <getopt.h>
+#include <unistd.h>
+#include <AK/Assertions.h>
+#include <LibCore/CFile.h>
+
+static void print_usage_and_exit()
+{
+    printf("usage: tail [-f, -n <linecount>] <filename>\n");
+    exit(1);
+}
+
+void tail_from_pos(CFile& file, off_t startline, bool want_follow)
+{
+    if (!file.seek(startline + 1))
+        return;
+
+    while (true) {
+        const auto& b = file.read(4096);
+        if (b.is_empty()) {
+            if (!want_follow) {
+                break;
+            } else {
+                while (!file.can_read()) {
+                    // FIXME: would be nice to have access to can_read_from_fd with an infinite timeout
+                    usleep(100);
+                }
+                continue;
+            }
+        }
+
+        write(STDOUT_FILENO, b.pointer(), b.size());
+    }
+}
+
+off_t find_seek_pos(CFile& file, int wanted_lines)
+{
+    // Rather than reading the whole file, start at the end and work backwards,
+    // stopping when we've found the number of lines we want.
+    off_t pos = 0;
+    if (!file.seek(0, CIODevice::SeekMode::FromEndPosition, &pos)) {
+        fprintf(stderr, "Failed to find end of file: %s\n", strerror(errno));
+        return -1;
+    }
+
+    off_t end = pos;
+    int lines = 0;
+
+    for(; pos >= 0; pos--) {
```

Comment:
Seeking one byte at a time seems hilariously inefficient, but if it works, it works!
Maybe add a FIXME about trying to seek one block at a time instead?

Reply:
True enough :)

Though, we could consider making CIODevice's read buffer deal with this internally too. Arguably, that'd be better than doing it independently in each application.